### DISCUSSION TITLE
How to configure reset on first logon but not reset on reboot?

### DISCUSSION THREAD

**Discussion by grojef** [quickfix-j/quickfixj#530] [Q&A]
How to configure reset on first logon but not reset on reboot?

**Comment by chrjohn** [comment]
What do you mean by reboot? Can you please show your config? A FIX sessions remains in existence between StartTime and EndTime or until you reset it.

**Reply by grojef** [reply]
If I configure resetOnLogon=Y, when I restart the application during transaction time, the application will logon again, and this time it will initiate a reset.
The transaction time needs to be restarted for some reason, but I don't want to reset it.
If reset, a resendRequest will be triggered, which will waste a lot of transaction time for recovery.

**Reply by chrjohn** [reply]
Then don't set ResetOnLogon? ;) The application should also reconnect when you don't reset the sequence numbers. That is the purpose of the sequence numbers. To continue the session where it was left off.

**Reply by grojef** [reply]
```
    private boolean generateLogon() throws IOException {
        final Message logon = messageFactory.create(sessionID.getBeginString(), MsgType.LOGON);
        logon.setInt(EncryptMethod.FIELD, 0);
        logon.setInt(HeartBtInt.FIELD, state.getHeartBeatInterval());
        if (sessionID.isFIXT()) {
            logon.setField(DefaultApplVerID.FIELD, senderDefaultApplVerID);
        }
        if (refreshOnLogon) {
            refreshState();
        }
        if (resetOnLogon) {
            resetState();
        }
        if (isResetNeeded()) {
            logon.setBoolean(ResetSeqNumFlag.FIELD, true);
        }
        state.setLastReceivedTime(SystemTime.currentTimeMillis());
        state.clearTestRequestCounter();
        state.setLogonSent(true);
        logonAttempts++;

        if (enableNextExpectedMsgSeqNum) {
            final int nextExpectedMsgNum = getExpectedTargetNum();
            logon.setInt(NextExpectedMsgSeqNum.FIELD, nextExpectedMsgNum);
            state.setLastExpectedLogonNextSeqNum(nextExpectedMsgNum);
        }

        setLogonTags(logon);
        return sendRaw(logon, 0);
    }
```
```
    private boolean isResetNeeded() {
        return sessionID.getBeginString().compareTo(FixVersions.BEGINSTRING_FIX41) >= 0
                && (resetOnLogon || resetOnLogout || resetOnDisconnect)
                && getExpectedSenderNum() == 1 && getExpectedTargetNum() == 1;
    }
```

I think ResetSeqNumFlag.FIELD is only needed for the first logon every day. If the transaction time is restarted for some reason, resetState() should not be triggered, if resetOnLogon || resetOnLogout || resetOnDisconnect are all set to N, then ResetSeqNumFlag.FIELD cannot be added when i first logon.

**Reply by chrjohn** [reply]
It will not be added if you don't set it. Please show your config.
Resetting the session will be done automatically based on StartTime and EndTime of the session as already said.

**Reply by grojef** [reply]
[default]
ConnectionType=initiator
SenderCompID=BANZI-UAT
TargetCompID=IB-UAT
StartTime=01:00:00
EndTime=23:00:00
HeartBtInt=30
ReconnectInterval=5
JdbcURL=jdbc:mysql://localhost:3306/fix?useSSL=false&useUnicode=true&characterEncoding=UTF8&allowMultiQueries=true&allowPublicKeyRetrieval=true&serverTimezone=Asia/Shanghai
JdbcUser=root
JdbcPassword=root123
JdbcDriver=com.mysql.cj.jdbc.Driver
[session]
BeginString=FIX.4.2
SocketConnectPort=9878
SocketConnectHost=127.0.0.1

**Reply by chrjohn** [reply]
Ok. Looks ok to me. Session reset will take place implicitly when you get either past EndTime or when you start your process after EndTime. It will detect that a new session has been started and reset implicitly, I.e. without you doing something or setting a ResetSeqNum flag. The counterparty of course needs to have the same schedule for this to work, I.e. they need to reset at the same time.

**Reply by grojef** [reply]
Yes, when endTime will be reset automatically, there is no need to set resetOnLogon || resetOnLogout || resetOnDisconnect to Y, but if the counterparty asks me to add ResetSeqNum flag when I logon for the first time, I must pass resetOnLogon || resetOnLogout || resetOnDisconnect One of them is set to Y to be added.

```
 private boolean isResetNeeded() {
        return sessionID.getBeginString().compareTo(FixVersions.BEGINSTRING_FIX41) >= 0
                && (resetOnLogon || resetOnLogout || resetOnDisconnect)
                && getExpectedSenderNum() == 1 && getExpectedTargetNum() == 1;
    }

```
My opinion is whether it is possible to add a resetOnFirstLogon to indicate the first logon reset, (resetOnLogon=N || resetOnLogout=N || resetOnDisconnect=N || resetOnFirstLogon=Y), so that you can logon for the first time without resetting the data Add ResetSeqNumFlag.FIELD, otherwise follow the previous logic.

**Reply by chrjohn** [reply]
That is not possible based on configuration. You would need to implement that as part of your application logic.
Your counter party should simply reset their sequence numbers based on a schedule and don't rely on you sending that flag only on the first logon.
